includes/Modules/Sign_In_With_Google.php

```diff
@@ -53,52 +66,208 @@ final class Sign_In_With_Google extends Module implements Module_With_Assets, Mo
+		if ( ! $user instanceof WP_User ) {
+			wp_safe_redirect( add_query_arg( 'error', self::INVALID_REQUEST_ERROR, $login_url ) );
 			exit;
 		}
+
+		// Redirect to the error page if the user is not a member of the current blog in multisite.
+		if ( is_multisite() ) {
+			$blog_id = get_current_blog_id();
+			if ( ! is_user_member_of_blog( $user->ID, $blog_id ) ) {
+				if ( get_option( 'users_can_register' ) ) {
+					add_user_to_blog( $blog_id, $user->ID, $this->get_default_role() );
+				} else {
+					wp_safe_redirect( add_query_arg( 'error', self::INVALID_REQUEST_ERROR, $login_url ) );
+					exit;
+				}
+			}
+		}
+
+		// Set the user to be the current user.
+		wp_set_current_user( $user->ID, $user->user_login );
+
+		// Set the authentication cookies and trigger the wp_login action.
+		wp_set_auth_cookie( $user->ID );
+		/** This filter is documented in wp-login.php */
+		do_action( 'wp_login', $user->user_login, $user );
+
+		// Use the admin dashboard URL as the redirect URL by default.
+		$redirect_to = admin_url();
+
+		// If we have the redirect URL in the cookie, use it as the main redirect_to URL.
+		$cookie_redirect_to = $this->context->input()->filter( INPUT_COOKIE, 'google_auth_redirect_to' );
+		if ( ! empty( $cookie_redirect_to ) ) {
+			$redirect_to = $cookie_redirect_to;
+			// phpcs:ignore WordPressVIPMinimum.Functions.RestrictedFunctions.cookies_setcookie
+			setcookie( 'google_auth_redirect_to', '', time() - 3600, $this->get_cookie_path(), COOKIE_DOMAIN );
+		}
+
+		// Redirect to HTTPS if user wants SSL.
+		if ( get_user_option( 'use_ssl', $user->ID ) && str_contains( $redirect_to, 'wp-admin' ) ) {
+			$redirect_to = preg_replace( '|^http://|', 'https://', $redirect_to );
+		}
+
+		/** This filter is documented in wp-login.php */
+		$redirect_to = apply_filters( 'login_redirect', $redirect_to, $redirect_to, $user );
+
+		if ( ( empty( $redirect_to ) || 'wp-admin/' === $redirect_to || admin_url() === $redirect_to ) ) {
+			// If the user doesn't belong to a blog, send them to user admin. If the user can't edit posts, send them to their profile.
+			if ( is_multisite() && ! get_active_blog_for_user( $user->ID ) && ! is_super_admin( $user->ID ) ) {
+				$redirect_to = user_admin_url();
+			} elseif ( is_multisite() && ! $user->has_cap( 'read' ) ) {
+				$redirect_to = get_dashboard_url( $user->ID );
+			} elseif ( ! $user->has_cap( 'edit_posts' ) ) {
+				$redirect_to = $user->has_cap( 'read' ) ? admin_url( 'profile.php' ) : home_url();
+			}
+		}
+
+		wp_safe_redirect( $redirect_to );
+		exit;
+	}
+
+	/**
+	 * Tries to find a user using user ID or email recieved from Google. If the user is not found,
+	 * attempts to create a new one.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return WP_User|WP_Error User object if found or created, WP_Error otherwise.
+	 */
+	private function find_or_create_user() {
+		$settings = $this->get_settings()->get();
+		$id_token = $this->context->input()->filter( INPUT_POST, 'credential' );
```

Comment:
Let's load all of the data/state from the request in the main handler (`handle_` method), and then pass it along to methods or other classes in parameters. This is a pattern we use in a few other places like `Setup` and `Authentication` which makes it a bit easier to test too.

Reply:
Updated.